### DISCUSSION TITLE
Rocket Chat Authentication Instructions?

### DISCUSSION THREAD

**Discussion by HiveMindNet** [requarks/wiki#3461] [Help / Questions]
**Question**
Trying to use the Rocket Chat auth but no documentation anywhere to help me test i. 

**Host Info (please complete the following information):**
OS: Latest Docker
Wiki.js version: 2.5.159
Database engine: Latest Docker PG_VERSION	11.8

Can someone please either point bee to some docs or help me understand where I get or set up the 'Client ID' and 'Client Secret' in my Rocket admin? Also, will this integration still work if I'm actually using iFrame auth on my Rocket instance? (Rocket authenticates from our other main website).

Any pointers or docs would be greatly appreciated :)

Thank you in advance.

**Comment by HiveMindNet** [comment]
Has anyone got this working at all?

**Comment by gatesry** [comment]
Have you looked at these docs https://docs.rocket.chat/guides/administrator-guides/authentication

Also check in the wiki code, it has an auth file that breaks down the strategy. I belive it's OAuth

**Comment by HiveMindNet** [comment]
Hi - Yes, that all refers to authenticating Rocket from another site. Not authenticating other sites from Rocket.

**Comment by HiveMindNet** [comment]
Has ***Anyone*** managed to get Rocket Chat Authentication working? If so could you please share how you set it up?

**Comment by erikvanberkum** [comment]
i got it installed to the point that i get the following error in wiki.js

Oops, something went wrong...
Missing or invalid email address from profile

![image](https://user-images.githubusercontent.com/7335299/107193974-cfac0b80-6a32-11eb-8a77-9948fa0973d8.png)

**Comment by deg0nz** [comment]
I have the exact same issue. 

I wanted to debug this further, but I can't find any method to increase the log level. I found no docs about some log level environment variable and mounting `config.yml` with increased `loglevel` seems not to work as mentioned in #2497.

What I found was, that the request to Wiki.js's auth callback url returns a HTTP 500.

**Comment by deg0nz** [comment]
Okay, I digged into this a little and found the following:

Wiki.js's oAuth module requests to Rocket.Chat API via [`/api/v1/me`](https://github.com/Requarks/wiki/blob/919d7c12a1c77d6e458f184586933e4212747676/server/modules/authentication/rocketchat/authentication.js#L16) and then extracts user information from the response. Wiki.js accesses [`usr.email`](https://github.com/Requarks/wiki/blob/919d7c12a1c77d6e458f184586933e4212747676/server/modules/authentication/rocketchat/authentication.js#L26) directly from the response object, but [according to the Rocket.Chat API docs](https://docs.rocket.chat/api/rest-api/methods/authentication/me), the user's emails are stored in an array called `emails`.

This could be the reason of the error because then the `email` field for the Wiki.js [callback function](https://github.com/Requarks/wiki/blob/919d7c12a1c77d6e458f184586933e4212747676/server/modules/authentication/rocketchat/authentication.js#L23) would be empty, since it does not exist in the response.

This should be a small fix. I will try to setup a dev environment for Wiki.js in the next days and send a PR if my findings are true.

Edit: wording

**Comment by deg0nz** [comment]
Also, @HiveMindNet's questions seem not to be answered yet.

> Can someone please either point bee to some docs or help me understand where I get or set up the 'Client ID' and 'Client Secret' in my Rocket admin? Also, will this integration still work if I'm actually using iFrame auth on my Rocket instance? (Rocket authenticates from our other main website).

`Client ID` and `Client Secret` are shown in Rocket.Chat's UI *after creating* the OAuth entry in Rocket.Chat admin UI.

I can't really answer the iFrame part, but there is an iFrame guide for Rocket.Chat. Maybe it helps: https://docs.rocket.chat/guides/developer/iframe-integration

**Comment by HiveMindNet** [comment]
Thanks @deg0nz - I don't need help with iFrame - that's all been working for ages; I just added for info purposes. The real question as you say still stands. Where to set or get the 'Client ID' and 'Client Secret' from.

**Comment by HiveMindNet** [comment]
On another forum, guys on there were saying that Rocket has no way of providing authentication services as a server. Which sounds like it contradicts the very purpose of this Wiki authentication strategy.

**Comment by deg0nz** [comment]
Okay, as I assumed, the problem was the `usr.email` property. I was able to authenticate with my Rocket.Chat instance and the fix I added in PR #3122. So Rocket.Chat as an OAuth server is definitely working.

I'm using Rocket.Chat 3.11.1

@HiveMindNet answers to your questions as follows:

> The real question as you say still stands. Where to set or get the 'Client ID' and 'Client Secret' from.

This is a little bit counter intuitive...

### Here is how you access the needed credentials in Rocket.Chat:

Create new OAuth Application in Rocket.Chat via Administration -> OAuth Apps

![Screenshot 2021-02-22 at 17 48 22](https://user-images.githubusercontent.com/6370295/108748015-6520c280-753e-11eb-9e01-bfca80d6f17d.png)

After that, you can access the created OAuth Application and read out `Client ID` and `Client Secret`:

![Screenshot 2021-02-22 at 18 12 17](https://user-images.githubusercontent.com/6370295/108748389-d496b200-753e-11eb-9b14-8b02d53c441b.png)
_Note: Don't forget to turn on your newly created OAuth application in Rocket.Chat! A newly added OAuth app is disabled by default. There is a small on/off switch at the top of the page._

Add the credentials to the corresponding fields in Wiki.js configuration.

Until PR #3122 is merged, you will get the error mentioned by @erikvanberkum though...

Edit: wording for more clarity and typos

**Comment by erikvanberkum** [comment]
The fix works well, I am able to login via Rocket.Chat 3.11.0.
Dont forget to set your user permissions in Wiki.JS

![image](https://user-images.githubusercontent.com/7335299/108792135-ee2e0d00-75c3-11eb-8d16-fcfae94ca4b8.png)

**Comment by HiveMindNet** [comment]
Thanks to @deg0nz and @erikvanberkum - I was so excited to see this progress today but then I tried and I don't think this has fixed the issue for me. I made the changes in the PR on Wiki.js but it's still failing with the following;

<img width="577" alt="Screenshot 2021-02-23 at 10 53 36" src="https://user-images.githubusercontent.com/19346725/108834920-a5c41e80-75c6-11eb-95bd-045005c01abd.png">

I'm not sure about what to expect if I simply add the rocket myaddress.com/oauth/token url to a browser - maybe it's expecting something or maybe that's my issue as I get 'Internal Server Error' on that URL :(

Does this look like it could be on the Rocket end now for me? Or are we still having problems with this Wiki auth?

I double checked all were enabled on rocket and wiki and it goes as far as having rocket ask to grant permission to Wiki.js - it's after I grant the permission that I get this fail to obtain the access token.

Again, any help is very much appreciated :)

**Reply by TBG-FR** [reply]
Reproduced it today. Glad to find I'm not the only one facing that issue 😅

**Comment by deg0nz** [comment]
So quick hint until PR #3122 is merged, here is a oneliner for a dockerized environment to fix the auth module in a running container:

``` bash
docker exec -it <CONTAINER_NAME> sed -i 's/email: usr.email/email: usr.emails[0].address/g' /wiki/server/modules/authentication/rocketchat/authentication.js
```

I automatically execute it after each container update

**Reply by HiveMindNet** [reply]
Thanks - it hasn't changed the result for me. Perhaps it needs a restart after the change?

**Reply by Phil-hostwiki** [reply]
Restart is required to effect changes.

**Reply by TBG-FR** [reply]
Thanks for your help @deg0nz ! However I executed your command, and double checked with a `cat | grep` the content of `/wiki/server/modules/authentication/rocketchat/authentication.js` and it looks right (`email: usr.emails[0].address,`), but I still have the error message "Failed to obtain access token", hence the restart 😞 

I'll keep digging !

**Reply by TBG-FR** [reply]
I tried to add some logs, I re-created the strategy in Wiki, generated a new app in RocketChat, restarted multiple times, and it's not getting better. I struggle to display the content of `usr`, the other debug/logs line I added are working, but not that one, **do you know if there is an access token verification prior to the rocketchat-specific code ?**

Oh and btw, when Wiki.JS starts, everything seems alright in logs : `Authentication Strategy RocketChat: [ OK ]`

**Reply by dudanogueira** [reply]
Hi!

There is probably a small "ahá!" here that may be easily missed (spent some time to discover this)

You must have the user before or, allow it to be created. Otherwise you can be stuck with what @HiveMindNet got.

Other then that, is working fine on latest versions on both side.
